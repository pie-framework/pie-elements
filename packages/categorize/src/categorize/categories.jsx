import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import { color } from '@pie-lib/pie-toolbox/render-ui';

import chunk from 'lodash/chunk';

import GridContent from './grid-content';
import Category, { CategoryType } from './category';

export { CategoryType };

export class Categories extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    categories: PropTypes.arrayOf(PropTypes.shape(CategoryType)),
    model: PropTypes.shape({
      categoriesPerRow: PropTypes.number,
    }),
    disabled: PropTypes.bool,
    onDropChoice: PropTypes.func.isRequired,
    onRemoveChoice: PropTypes.func.isRequired,
    rowLabels: PropTypes.array
  };

  static defaultProps = {
    model: {
      categoriesPerRow: 1,
    },
  };

  render() {
    const { classes, categories, model, disabled, onDropChoice, onRemoveChoice, rowLabels } = this.props;
    const { categoriesPerRow, minRowHeight } = model;

    // split categories into an array of arrays (inner array),
    // where each inner array represents how many categories should be displayed on one row
    const chunkedCategories = chunk(categories, categoriesPerRow);

    const hasNonEmptyString = (array) => {
      let found = false;

      (array || []).forEach(element => {
        if (typeof element === 'string' && element.trim() !== '' && element.trim() !== '<div></div>') {
          found = true;
        }
      });

      return found;
    };

    return (
      <GridContent
        columns={categoriesPerRow}
        className={classes.categories}
        rows={Math.ceil(categories.length / categoriesPerRow) * 2}
      >
        {chunkedCategories.map((cat, rowIndex) => {
          let items = [];

          // for each inner array of categories, create a row with category titles
          // first cell of row has to be the row label
          cat.forEach((c, columnIndex) => {
            items.push(
              <div style={{ display: 'flex' }}>
                {columnIndex === 0 && hasNonEmptyString(rowLabels) ? (
                  <div
                    key={rowIndex}
                    className={classes.rowLabel}
                    dangerouslySetInnerHTML={{
                      __html: rowLabels[rowIndex] || '',
                    }}
                  />
                ) : null}
                <div className={classes.categoryWrapper}>
                  <div
                    className={classes.label}
                    key={`category-label-${rowIndex}-${columnIndex}`}
                    dangerouslySetInnerHTML={{ __html: c.label }}
                  />

                  <Category
                    minRowHeight={minRowHeight}
                    onDropChoice={(h) => onDropChoice(c.id, h)}
                    onRemoveChoice={onRemoveChoice}
                    disabled={disabled}
                    className={classes.category}
                    key={`category-element-${rowIndex}-${columnIndex}`}
                    {...c}
                  />
                </div>
              </div>,
            );
          });

          // if the last row has fewer categories than max on a row, fill the spaces with divs
          items = items.concat(
            Array(categoriesPerRow - cat.length)
              .fill(<div />)
              .map((value, index) => <div key={`fill-space-final-${index}`} />),
          );

          return items;
        })}
      </GridContent>
    );
  }
}

const styles = (theme) => ({
  categories: {
    flex: 1,
  },
  label: {
    color: color.text(),
    backgroundColor: color.background(),
    textAlign: 'center',
    paddingTop: theme.spacing.unit,
  },
  rowLabel: {
    alignItems: 'center',
    display: 'flex',
    justifyContent: 'center',
    flex: 0.5,
    marginRight: '12px',
  },
  categoryWrapper: {
    display: 'flex',
    flex: '2',
    flexDirection: 'column',
  },
});
export default withStyles(styles)(Categories);
