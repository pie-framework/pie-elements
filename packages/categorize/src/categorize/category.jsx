import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import Choice from './choice';
import PlaceHolder from './droppable-placeholder';
import { color } from '@pie-lib/render-ui';

export const CategoryType = {
  id: PropTypes.string.isRequired,
  categoryId: PropTypes.string,
};

export class Category extends React.Component {
  static propTypes = {
    ...CategoryType,
    className: PropTypes.string,
    disabled: PropTypes.bool,
    classes: PropTypes.object.isRequired,
    onDropChoice: PropTypes.func,
    onRemoveChoice: PropTypes.func,
  };

  static defaultProps = {};

  render() {
    const {
      classes,
      className,
      choices = [],
      disabled,
      onDropChoice,
      onRemoveChoice,
      id,
      correct,
    } = this.props;

    const names = classNames(classes.category, className);
    const placeholderNames = classNames(
      classes.placeholder,
      correct === false && classes.incorrect
    );

    return (
      <div className={names}>
        <PlaceHolder
          onDropChoice={onDropChoice}
          disabled={disabled}
          className={placeholderNames}
        >
          {choices.map((c, index) => (
            <Choice
              onRemoveChoice={onRemoveChoice}
              disabled={disabled}
              key={index}
              choiceIndex={index}
              categoryId={id}
              {...c}
            />
          ))}
        </PlaceHolder>
      </div>
    );
  }
}

const styles = (theme) => ({
  incorrect: {
    border: `solid 2px ${color.incorrect()}`,
  },
  placeholder: {
    minHeight: '80px',
    padding: theme.spacing.unit / 2,
    borderRadius: theme.spacing.unit / 2,
    gridColumnGap: 0,
    gridRowGap: 0,
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'center',
    alignItems: 'center',
    alignContent: 'flex-start'
  },
  category: {
    display: 'flex',
    flexDirection: 'column',
  },
});

export default withStyles(styles)(Category);
