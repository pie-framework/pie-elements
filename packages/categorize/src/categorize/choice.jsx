import React from 'react';
import PropTypes from 'prop-types';
import { useDrag } from 'react-dnd';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import { color } from '@pie-lib/pie-toolbox/render-ui';

export const ChoiceType = {
  content: PropTypes.string.isRequired,
  id: PropTypes.string,
};

export const Layout = ({ classes, className, content, isDragging, disabled, correct }) => {
  const rootNames = classNames(
      correct === true && 'correct',
      correct === false && 'incorrect',
      classes.choice,
      isDragging && classes.dragging,
      disabled && classes.disabled,
      className,
  );
  const cardNames = classNames(classes.card);
  return (
      <div className={rootNames}>
        <Card className={cardNames}>
          <CardContent
              classes={{ root: classes.cardRoot }}
              dangerouslySetInnerHTML={{ __html: content }}
          />
        </Card>
      </div>
  );
};

Layout.propTypes = {
  ...ChoiceType,
  classes: PropTypes.object.isRequired,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  correct: PropTypes.bool,
};

const styles = (theme) => ({
  choice: {
    direction: 'initial',
    cursor: 'pointer',
    width: '100%',
    '&.correct': {
      border: `solid 2px ${color.correct()}`,
    },
    '&.incorrect': {
      border: `solid 2px ${color.incorrect()}`,
    },
    borderRadius: '6px',
  },
  cardRoot: {
    color: color.text(),
    backgroundColor: color.white(),
    fontSize: theme.typography.fontSize + 2,
    '&:last-child': {
      paddingBottom: theme.spacing.unit * 2,
    },
    borderRadius: '4px',
    border: '1px solid',
  },
  disabled: {
    cursor: 'not-allowed',
    opacity: '0.6',
  },
  dragging: {
    cursor: 'move',
  },
  card: {
    color: color.text(),
    backgroundColor: color.background(),
    width: '100%',
    pointerEvents: 'none',
  },
});

const Styled = withStyles(styles)(Layout);

const Choice = ({ id, content, disabled, correct, extraStyle, onRemoveChoice }) => {
  const [{ isDragging }, drag] = useDrag({
    type: 'categorize',
    item: { id, content },
    canDrag: !disabled,
    end: (item, monitor) => {
      if (!monitor.didDrop()) {
        onRemoveChoice(item);
      }
    },
    collect: (monitor) => ({
      isDragging: monitor.isDragging(),
    }),
  });

  return (
      <div ref={drag} style={{ margin: '4px', ...extraStyle }}>
        <Styled
            id={id}
            content={content}
            disabled={disabled}
            correct={correct}
            isDragging={isDragging}
        />
      </div>
  );
};

Choice.propTypes = {
  ...ChoiceType,
  extraStyle: PropTypes.object,
  onRemoveChoice: PropTypes.func.isRequired,
};

export default Choice;
