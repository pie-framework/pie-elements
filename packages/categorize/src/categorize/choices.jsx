import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Choice, { ChoiceType } from './choice';
export { ChoiceType };

const Blank = () => <div />;

export class Choices extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    choices: PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.shape(ChoiceType),
        PropTypes.shape({ empty: PropTypes.bool }),
      ])
    ),
    model: PropTypes.shape({
      categoriesPerRow: PropTypes.number,
      choicesLabel: PropTypes.string,
    }),
    disabled: PropTypes.bool,
    choicePosition: PropTypes.string,
  };

  static defaultProps = {
    model: {
      categoriesPerRow: 1,
      choicesLabel: '',
    },
  };

  render() {
    const {
      classes,
      choices = [],
      model,
      disabled,
      choicePosition,
    } = this.props;
    let style = {
      textAlign: 'center',
    };

    if (choicePosition === 'left') {
      style.direction = 'rtl';
    }

    return (
      <div className={classes.wrapper}>
        {model.choicesLabel && model.choicesLabel !== '' && (
          <div className={classes.labelHolder}>{model.choicesLabel}</div>
        )}
        <div className={classes.choices} style={style}>
          {choices.map((c, index) => {
            return c.empty ? (
              <Blank key={index} />
            ) : (
              <Choice
                disabled={disabled}
                className={classes.choice}
                key={index}
                extraStyle={{ maxWidth: `${95 / model.categoriesPerRow}%` }}
                {...c}
              />
            );
          })}
        </div>
      </div>
    );
  }
}

const styles = (theme) => ({
  wrapper: {
    flex: 1,
    padding: theme.spacing.unit,
  },
  choices: {
    padding: theme.spacing.unit / 2,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    flexWrap: 'wrap'
  },
  labelHolder: {
    margin: '0 auto',
    textAlign: 'center',
    paddingTop: theme.spacing.unit,
  },
});

export default withStyles(styles)(Choices);
