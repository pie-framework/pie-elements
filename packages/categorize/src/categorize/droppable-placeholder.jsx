import React from 'react';
import { PlaceHolder } from '@pie-lib/pie-toolbox/drag';
import PropTypes from 'prop-types';
import { DropTarget } from 'react-dnd';
import { uid } from '@pie-lib/pie-toolbox/drag';
import debug from 'debug';

const log = debug('@pie-ui:categorize:droppable-placeholder');

export class DroppablePlaceholder extends React.Component {
  static propTypes = {
    choiceBoard: PropTypes.bool,
    connectDropTarget: PropTypes.func.isRequired,
    isOver: PropTypes.bool,
    children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
    className: PropTypes.string,
    grid: PropTypes.object,
    disabled: PropTypes.bool,
    minRowHeight: PropTypes.string,
  };
  render() {
    const {
      children,
      connectDropTarget,
      isOver,
      className,
      grid,
      disabled,
      choiceBoard,
      minRowHeight
    } = this.props;

    return connectDropTarget(
      <div style={{ flex: 1, minHeight: minRowHeight || '10px' }}>
        <PlaceHolder
          className={className}
          isOver={isOver}
          grid={grid}
          disabled={disabled}
          choiceBoard={choiceBoard}
          isCategorize
        >
          {children}
        </PlaceHolder>
      </div>,
    );
  }
}

export const spec = {
  drop: (props, monitor) => {
    log('[drop] props: ', props);
    const item = monitor.getItem();
    props.onDropChoice(item);
  },
  canDrop: (props /*, monitor*/) => {
    return !props.disabled;
  },
};

const WithTarget = DropTarget(
  ({ uid }) => uid,
  spec,
  (connect, monitor) => ({
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver(),
  }),
)(DroppablePlaceholder);

export default uid.withUid(WithTarget);
