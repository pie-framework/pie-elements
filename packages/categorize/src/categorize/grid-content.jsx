import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';

export class GridContent extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
    columns: PropTypes.number,
    rows: PropTypes.number,
    extraStyle: PropTypes.object,
  };
  static defaultProps = {
    columns: 2,
    rows: 2,
  };

  render() {
    const { classes, className, children, columns, extraStyle, rows } = this.props;
    const style = {
      gridTemplateColumns: `repeat(${columns}, 1fr)`,
      gridTemplateRows: rows === 2 ? 'auto 1fr' : `repeat(${rows}, auto)`,
      ...extraStyle,
    };

    return (
      <div style={style} className={classNames(classes.gridContent, className)}>
        {children}
      </div>
    );
  }
}

const styles = (theme) => ({
  gridContent: {
    display: 'grid',
    columnGap: `${theme.spacing.unit}px`,
    gridColumnGap: `${theme.spacing.unit}px`,
    rowGap: `${theme.spacing.unit}px`,
    gridRowGap: `${theme.spacing.unit}px`,
    gridAutoRows: '1fr',
  },
});

export default withStyles(styles)(GridContent);
