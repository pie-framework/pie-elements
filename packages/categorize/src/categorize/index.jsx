import React from 'react';
import PropTypes from 'prop-types';
import Choices from './choices';
import Categories from './categories';
import CorrectAnswerToggle from '@pie-lib/correct-answer-toggle';
import { withStyles } from '@material-ui/core/styles';
import {
  buildState,
  removeChoiceFromCategory,
  moveChoiceToCategory
} from '@pie-lib/categorize';
import { withDragContext, uid } from '@pie-lib/drag';
import { color, Feedback, Collapsible, hasText } from '@pie-lib/render-ui';
import debug from 'debug';

const log = debug('@pie-ui:categorize');

const removeHTMLTags = html => {
  const tmp = document.createElement('DIV');

  tmp.innerHTML = html;

  const value = tmp.textContent || tmp.innerText || '';

  return value.trim();
};

export class Categorize extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    model: PropTypes.object,
    session: PropTypes.shape({
      answers: PropTypes.arrayOf(
        PropTypes.shape({
          choice: PropTypes.string,
          category: PropTypes.string
        })
      )
    }),
    onAnswersChange: PropTypes.func.isRequired,
    onShowCorrectToggle: PropTypes.func.isRequired
  };

  static defaultProps = {
    disabled: false
  };

  constructor(props) {
    super(props);

    this.state = {
      showCorrect: false
    };
  }

  dropChoice = (categoryId, draggedChoice) => {
    const { session, onAnswersChange } = this.props;
    log(
      '[dropChoice] category: ',
      draggedChoice.categoryId,
      'choice: ',
      draggedChoice
    );

    const answers = moveChoiceToCategory(
      draggedChoice.id,
      draggedChoice.categoryId,
      categoryId,
      draggedChoice.choiceIndex,
      session.answers
    );

    onAnswersChange(answers);
  };

  removeChoice = c => {
    log('[removeChoice]: ', c);
    const { onAnswersChange, session } = this.props;
    const answers = removeChoiceFromCategory(
      c.id,
      c.categoryId,
      c.choiceIndex,
      session.answers
    );
    onAnswersChange(answers);
  };

  UNSAFE_componentWillReceiveProps() {
    this.setState({ showCorrect: false });
  }

  toggleShowCorrect = () =>
    this.setState({ showCorrect: !this.state.showCorrect }, () => {
      this.props.onShowCorrectToggle();
    });

  getPositionDirection = (choicePosition) => {
    let flexDirection;

    switch (choicePosition) {
      case 'left':
        flexDirection = 'row-reverse';
        break;
      case 'right':
        flexDirection = 'row';
        break;
      case 'below':
        flexDirection = 'column';
        break;
      default:
        // above
        flexDirection = 'column-reverse';
        break;
    }

    return flexDirection;
  };

  render() {
    const { classes, model, session } = this.props;
    const { showCorrect } = this.state;
    const { choicesPosition, note, showNote, env } = model;
    const { mode, role } = env || {};
    const choicePosition = choicesPosition || 'above';

    const style = {
      flexDirection: this.getPositionDirection(choicePosition)
    };

    const { categories, choices, correct } = buildState(
      model.categories,
      model.choices,
      showCorrect ? model.correctResponse : session.answers,
      model.correctResponse
    );

    log('[render] disabled: ', model.disabled);

    const { rowLabels, categoriesPerRow } = model;
    const nbOfRows = categories && Math.ceil(categories.length / categoriesPerRow) || 0;
    const displayNote = (showCorrect || mode === 'view' && role === 'instructor') && showNote && note;

    return (
      <div className={classes.mainContainer}>
        {
          model.teacherInstructions && hasText(model.teacherInstructions) && (
            <React.Fragment>
              <Collapsible
                labels={{ hidden: 'Show Teacher Instructions', visible: 'Hide Teacher Instructions' }}
                className={classes.collapsible}
              >
                <div dangerouslySetInnerHTML={{ __html: model.teacherInstructions }}/>
              </Collapsible>
              <br />
            </React.Fragment>
          )
        }
        <CorrectAnswerToggle
          show={showCorrect || correct === false}
          toggled={showCorrect}
          onToggle={this.toggleShowCorrect}
        />
        {
          model.prompt && removeHTMLTags(model.prompt) &&
          <div
            className={classes.prompt}
            dangerouslySetInnerHTML={{ __html: model.prompt }}
          />
        }
        <div className={classes.categorize} style={style}>
          <div style={{ display: 'flex' }}>
            {
              rowLabels && nbOfRows && (
                <div style={{ display: 'grid', marginRight: '20px' }}>
                  {rowLabels.slice(0, nbOfRows).map((label, index) => (
                    <div
                      key={index}
                      style={{
                        alignItems: 'center',
                        display: 'flex',
                        justifyContent: 'center'
                      }}
                      dangerouslySetInnerHTML={{
                        __html: label
                      }}
                    />
                  ))}
                </div>
              )
            }
            <Categories
              model={model}
              disabled={model.disabled}
              categories={categories}
              onDropChoice={this.dropChoice}
              onRemoveChoice={this.removeChoice}
            />
          </div>
          <Choices
            disabled={model.disabled}
            model={model}
            choices={choices}
            choicePosition={choicePosition}
          />
        </div>
        {displayNote && (
          <div
            className={classes.note}
            dangerouslySetInnerHTML={{ __html: `<strong>Note:</strong> ${note}` }}
          />
        )}
        {
          model.rationale && hasText(model.rationale) && (
            <Collapsible
              labels={{ hidden: 'Show Rationale', visible: 'Hide Rationale' }}
              className={classes.collapsible}
            >
              <div dangerouslySetInnerHTML={{ __html: model.rationale }}/>
            </Collapsible>
          )
        }
        {
          model.correctness &&
          model.feedback &&
          !showCorrect && (
            <Feedback
              correctness={model.correctness}
              feedback={model.feedback}
            />
          )
        }
      </div>
    );
  }
}

class CategorizeProvider extends React.Component {
  constructor(props) {
    super(props);
    this.uid = uid.generateId();
  }

  render() {
    return (
      <uid.Provider value={this.uid}>
        <Categorize { ...this.props } />
      </uid.Provider>
    );
  }
}

const styles = (theme) => ({
  mainContainer: {
    padding: theme.spacing.unit,
    color: color.text(),
    backgroundColor: color.background()
  },
  prompt: {
    marginBottom: '35px',
    verticalAlign: 'middle'
  },
  note: {
    padding: '5px 0'
  },
  categorize: {
    marginBottom: theme.spacing.unit,
    display: 'flex',
    flexDirection: 'column'
  },
  collapsible: {
    paddingBottom: theme.spacing.unit * 2
  }
});
export default withDragContext(withStyles(styles)(CategorizeProvider));
