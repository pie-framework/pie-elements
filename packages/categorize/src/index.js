import React from 'react';
import ReactDOM from 'react-dom';
import { renderMath } from '@pie-lib/pie-toolbox/math-rendering';
import { SessionChangedEvent, ModelSetEvent } from '@pie-framework/pie-player-events';
import CategorizeComponent from './categorize';

export default class Categorize extends HTMLElement {
  set model(m) {
    this._model = m;

    this.eliminateBlindAnswersFromSession();
    this.dispatchEvent(new ModelSetEvent(this.tagName.toLowerCase(), this.isComplete(), !!this._model));
    this.render();
  }

  isComplete() {
    if (!this._session || !this._session.answers) {
      return false;
    }

    if (!Array.isArray(this._session.answers)) {
      return false;
    }

    return this._session.answers.some((answer) => answer.choices && answer.choices.length > 0);
  }

  set session(s) {
    if (s && !s.answers) {
      s.answers = [];
    }

    this._session = s;
    this.render();
  }

  get session() {
    return this._session;
  }

  eliminateBlindAnswersFromSession() {
    const { answers = [] } = this._session || {};
    const { choices = [] } = this._model || {};

    const mappedChoices = choices.map((c) => c.id) || [];
    const filteredAnswers = answers.map((answer) => {
      const answerChoices = answer?.choices || [];
      answer.choices = answerChoices.filter((c) => mappedChoices.includes(c));

      return answer;
    });

    if (filteredAnswers.length > 0) {
      this.changeAnswers(filteredAnswers);
    }
  }

  changeAnswers(answers) {
    this._session.answers = answers;

    this.dispatchEvent(new SessionChangedEvent(this.tagName.toLowerCase(), this.isComplete()));

    this.render();
  }

  onShowCorrectToggle() {
    renderMath(this);
  }

  render() {
    if (this._model && this._session) {
      const el = React.createElement(CategorizeComponent, {
        model: this._model,
        session: this._session,
        onAnswersChange: this.changeAnswers.bind(this),
        onShowCorrectToggle: this.onShowCorrectToggle.bind(this),
      });

      ReactDOM.render(el, this, () => {
        renderMath(this);
      });
    }
  }
}
