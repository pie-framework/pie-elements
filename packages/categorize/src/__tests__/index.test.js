import React from 'react';
import Categorize from '../index';
import { shallow } from 'enzyme';
import { ModelSetEvent, SessionChangedEvent } from '@pie-framework/pie-player-events';
import { Categorize as UnStyledCategorize } from '../categorize/index';

jest.mock('@pie-lib/math-rendering', () => ({ renderMath: jest.fn() }));

describe('categorize', () => {
  describe('renders', () => {
    let wrapper = (props) => {
      let defaultProps = {
        model: {
          categories: [],
          choices: [],
          correctResponse: [],
          ...props,
        },
        session: {},
        classes: {},
      };
      return shallow(<UnStyledCategorize {...defaultProps} />);
    };

    it('snapshot', () => {
      const w = wrapper();
      expect(w).toMatchSnapshot();
    });

    it('snapshot with rationale', () => {
      const w = wrapper({ rationale: 'This is rationale' });
      expect(w).toMatchSnapshot();
    });

    it('snapshot with teacherInstructions', () => {
      const w = wrapper({ teacherInstructions: 'These are teacher instructions' });
      expect(w).toMatchSnapshot();
    });
  });

  describe('events', () => {
    describe('model', () => {
      it('dispatches model set event', () => {
        const el = new Categorize();
        el.tagName = 'categorize-el';
        el.model = {};
        expect(el.dispatchEvent).toBeCalledWith(new ModelSetEvent('categorize-el', false, true));
      });
    });

    describe('changeAnswers', () => {
      it('dispatches session changed event - add answer', () => {
        const el = new Categorize();
        el.tagName = 'categorize-el';
        el.session = { answers: [] };
        el.changeAnswers([{ category: 'id-fruits', choices: ['apple'] }]);
        expect(el.dispatchEvent).toBeCalledWith(new SessionChangedEvent('categorize-el', true));
      });

      it('dispatches session changed event - remove answer', () => {
        const el = new Categorize();
        el.tagName = 'categorize-el';
        el.session = { answers: [{ category: 'id-fruits', choices: ['apple'] }] };
        el.changeAnswers([{ category: 'id-fruits', choices: [] }]);
        expect(el.dispatchEvent).toBeCalledWith(new SessionChangedEvent('categorize-el', false));
      });

      it('dispatches session changed event - add/remove answer', () => {
        const el = new Categorize();
        el.tagName = 'categorize-el';
        el.session = { answers: [{ category: 'id-fruits', choices: ['apple'] }] };
        el.changeAnswers([
          { category: 'id-fruits', choices: ['apple'] },
          { category: 'id-vegetables', choices: ['carrot'] },
        ]);
        expect(el.dispatchEvent).toBeCalledWith(new SessionChangedEvent('categorize-el', true));

        el.changeAnswers([
          { category: 'id-fruits', choices: ['apple'] },
          { category: 'id-vegetables', choices: [] },
        ]);
        expect(el.dispatchEvent).toBeCalledWith(new SessionChangedEvent('categorize-el', true));

        el.changeAnswers([
          { category: 'id-fruits', choices: [] },
          { category: 'id-vegetables', choices: [] },
        ]);
        expect(el.dispatchEvent).toBeCalledWith(new SessionChangedEvent('categorize-el', false));
      });
    });
  });
});
