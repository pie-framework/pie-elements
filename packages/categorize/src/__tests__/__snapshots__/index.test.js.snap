// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`categorize renders snapshot 1`] = `
<div>
  <WithStyles(CorrectAnswerToggle)
    onToggle={[Function]}
    show={false}
    toggled={false}
  />
  <div
    style={
      Object {
        "flexDirection": "column-reverse",
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <WithStyles(Categories)
        categories={Array []}
        model={
          Object {
            "categories": Array [],
            "choices": Array [],
            "correctResponse": Array [],
          }
        }
        onDropChoice={[Function]}
        onRemoveChoice={[Function]}
      />
    </div>
    <WithStyles(Choices)
      choicePosition="above"
      choices={Array []}
      model={
        Object {
          "categories": Array [],
          "choices": Array [],
          "correctResponse": Array [],
        }
      }
    />
  </div>
</div>
`;

exports[`categorize renders snapshot with rationale 1`] = `
<div>
  <WithStyles(CorrectAnswerToggle)
    onToggle={[Function]}
    show={false}
    toggled={false}
  />
  <div
    style={
      Object {
        "flexDirection": "column-reverse",
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <WithStyles(Categories)
        categories={Array []}
        model={
          Object {
            "categories": Array [],
            "choices": Array [],
            "correctResponse": Array [],
            "rationale": "This is rationale",
          }
        }
        onDropChoice={[Function]}
        onRemoveChoice={[Function]}
      />
    </div>
    <WithStyles(Choices)
      choicePosition="above"
      choices={Array []}
      model={
        Object {
          "categories": Array [],
          "choices": Array [],
          "correctResponse": Array [],
          "rationale": "This is rationale",
        }
      }
    />
  </div>
  <WithStyles(Collapsible)
    labels={
      Object {
        "hidden": "Show Rationale",
        "visible": "Hide Rationale",
      }
    }
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "This is rationale",
        }
      }
    />
  </WithStyles(Collapsible)>
</div>
`;

exports[`categorize renders snapshot with teacherInstructions 1`] = `
<div>
  <WithStyles(Collapsible)
    labels={
      Object {
        "hidden": "Show Teacher Instructions",
        "visible": "Hide Teacher Instructions",
      }
    }
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "These are teacher instructions",
        }
      }
    />
  </WithStyles(Collapsible)>
  <br />
  <WithStyles(CorrectAnswerToggle)
    onToggle={[Function]}
    show={false}
    toggled={false}
  />
  <div
    style={
      Object {
        "flexDirection": "column-reverse",
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <WithStyles(Categories)
        categories={Array []}
        model={
          Object {
            "categories": Array [],
            "choices": Array [],
            "correctResponse": Array [],
            "teacherInstructions": "These are teacher instructions",
          }
        }
        onDropChoice={[Function]}
        onRemoveChoice={[Function]}
      />
    </div>
    <WithStyles(Choices)
      choicePosition="above"
      choices={Array []}
      model={
        Object {
          "categories": Array [],
          "choices": Array [],
          "correctResponse": Array [],
          "teacherInstructions": "These are teacher instructions",
        }
      }
    />
  </div>
</div>
`;
